src/sinks/util/buffer/metrics.rs

```diff
@@ -281,6 +197,102 @@ impl Batch for MetricBuffer {
     }
 }
 
+/// The metrics state trait abstracts how data point normalization is
+/// done.  Normalisation is required to make sure Sources and Sinks are
+/// exchanging compatible data structures. For instance, delta gauges
+/// produced by Statsd source cannot be directly sent to Datadog API. In
+/// this case the buffer will keep the state of a gauge value, and
+/// produce absolute values gauges that are well supported by Datadog.
+///
+/// Another example of normalisation is disaggregation of counters. Most
+/// sinks would expect we send them delta counters (e.g. how many events
+/// occurred during the flush period). And most sources are producing
+/// exactly these kind of counters, with Prometheus being a notable
+/// exception. If the counter comes already aggregated inside the
+/// source, the buffer will compare it's values with the previous known
+/// and calculate the delta.
+pub trait MetricNormalize {
+    fn apply_state(state: &mut MetricSet, metric: Metric) -> Option<Metric>;
+}
+
+/// This is a convenience newtype wrapper for HashSet<MetricEntry> that
+/// provides some extra functionality.
+#[derive(Clone, Default)]
+pub struct MetricSet(HashSet<MetricEntry>);
+
+impl Deref for MetricSet {
+    type Target = HashSet<MetricEntry>;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl DerefMut for MetricSet {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl MetricSet {
+    fn with_capacity(capacity: usize) -> Self {
+        Self(HashSet::with_capacity(capacity))
+    }
+
+    /// Either pass the metric through as-is if absolute, or convert it
+    /// to absolute if incremental.
+    pub fn make_absolute(&mut self, metric: Metric) -> Option<Metric> {
+        match metric.data.kind {
+            MetricKind::Absolute => Some(metric),
+            MetricKind::Incremental => self.incremental_to_absolute(metric),
+        }
+    }
+
+    /// Either convert the metric to incremental if absolute, or
+    /// aggregate it with any previous value if already incremental.
+    pub fn make_incremental(&mut self, metric: Metric) -> Option<Metric> {
+        match metric.data.kind {
+            MetricKind::Absolute => self.absolute_to_incremental(metric),
+            MetricKind::Incremental => Some(metric),
+        }
+    }
+
+    /// Convert the incremental metric into an absolute one, using the
+    /// state buffer to keep track of the value throughout the entire
+    /// application uptime.
+    fn incremental_to_absolute(&mut self, metric: Metric) -> Option<Metric> {
+        let mut entry = MetricEntry(metric.into_absolute());
+        let mut existing = self.0.take(&entry).unwrap_or_else(|| {
+            // Start from zero value if the entry is not found.
+            MetricEntry(entry.zero())
+        });
+        existing.data.value.add(&entry.data.value);
+        entry.data.value = existing.data.value.clone();
+        self.0.insert(existing);
```

Comment:
Would the following be the same?

```suggestion
        entry.data.value.add(&existing.data.value);
        self.0.insert(entry);
```

Reply:
Nope, because `insert` takes ownership of `entry`, and it is returned in the next line.